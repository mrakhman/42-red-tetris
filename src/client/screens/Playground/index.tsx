import  React, { useState, useEffect } from 'react';
import _ from 'lodash';
import { View, Text } from 'react-native';
import useInterval from '@use-it/interval';
import { useRoute, RouteProp } from '@react-navigation/native';
import io from 'socket.io-client';

import { blankMatrix, blockMatrix } from '/client/constants/tetriminos';
import { blockTypes } from '/client/constants/tetriminos';
import { ChatWidget, addResponseMessage } from '/client/components/Chat';
import Block from '/client/models/block';
import Digits from '/client/components/Digits';
import formatChatSubtitle from '/client/screens/Playground/utils';
import Gameboy from '/client/components/Gameboy';
import Matrix from '/client/components/Matrix';
import { useKeyEvent } from '/client/hooks/useKeyEvent';

export default function Playground(): JSX.Element {
  const route = useRoute<RouteProp<RootStackParamList, 'Playground'>>();
  const { params } = route;
  const { room, username } = params ?? {};
  const [roomUsers, setRoomUsers] = useState<string[]>([]);
  const [block, setBlock] = useState<Block>(new Block({ type: _.sample(blockTypes) ?? 'T' }));
  const [matrix, setMatrix] = useState<Matrix>(blankMatrix);
  const [isPause, setIsPause] = useState<boolean>(true);
  useKeyEvent({ setBlock, setMatrix, setIsPause });

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let socket: any = null;

  // Initialize new socket
  useEffect(() => {
    socket = io(`${process.env.EXPO_SOCKET_URL}`);
  }, []);

  useEffect(() => {
    socket.emit('joinRoom', { username, roomName: room });
    // Message from server
    socket.on('message', (message: Message) => {
      if (message.username !== username)
        addResponseMessage(message.username + ': ' + message.text, message.username);
    });
    socket.on('update room users', (data: { room: string, users: UserType[] }) => {
      setRoomUsers(data.users.map((user) => user.username));
    });
  }, []);
  const handleNewUserMessage = (message: string) => {
    socket.emit('chatMessage', { username, message, room });
  };
  const nextBlockType = blockTypes[(_.indexOf(blockTypes, block.type) + 1) % _.size(blockTypes)];
  useInterval(() => {
    if (isPause) return;
    if (_.includes(matrix[0], 1)) {
      setMatrix(blankMatrix);
      setIsPause(true);
    } else {
      setBlock((currentBlock) => {
        if (!block.fall().isValid(matrix)) {
          setMatrix(block.printBlock(matrix));
          return new Block({ type: nextBlockType });
        } else {
          return currentBlock.fall();
        }
      });
    }
  }, 500);

  return (
    <>
      <Gameboy isPause={isPause}>
        <View style={{ flexDirection: 'row', alignSelf:'flex-start', width: '100%' }}>
          <Matrix matrix={matrix} block={block}/>
          <View style={{ marginLeft: 20, flex: 1 }} >
            <Text style={{ fontSize: 20 }}>Score</Text>
            <Digits style={{ marginVertical: 10, alignSelf: 'flex-end' }} />
            <Text style={{ fontSize: 20 }}>Start Line</Text>
            <Digits style={{ marginVertical: 10, alignSelf: 'flex-end' }} />
            <Text style={{ fontSize: 20 }}>Level</Text>
            <Digits style={{ marginVertical: 10, alignSelf: 'flex-end' }} />
            <Text style={{ fontSize: 20 }}>Next</Text>
            <Matrix
              matrix={blockMatrix}
              block={new Block({ type: nextBlockType, pos: [0, 0] })}
              style={{ borderWidth: 0, marginVertical: 10, alignSelf: 'flex-end' }}
            />
          </View>
        </View>
      </Gameboy>
      {room && username &&
        <ChatWidget title="RedTetris" subtitle={formatChatSubtitle(roomUsers)} handleNewUserMessage={handleNewUserMessage}/>
      }
    </>
  );
}
