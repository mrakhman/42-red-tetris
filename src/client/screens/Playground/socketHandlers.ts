import _ from 'lodash';
import { Dispatch, SetStateAction } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';

import { SOCKETS, SCORING } from '/config/constants';
import { addResponseMessage } from '/client/components/Chat';
import { penaltyLine, blankMatrix } from '/client/constants/tetriminos';

const addPenaltyRows = (matrix: Matrix, rowsNumber: number): Matrix => {
  // Create array of blank lines
  const penaltyMatrix = Array(rowsNumber).fill(penaltyLine);
  // Add it to the bottom of matrix (matrix has 20 + rowsNumber lines)
  const newMatrix = _.cloneDeep([...matrix, ...penaltyMatrix]);
  // Return matrix without N start lines (matrix has 20 lines again)
  return _.slice(newMatrix, rowsNumber, newMatrix.length);
};

/*
** socket.emit eveins:
*/

/* SOCKETS.CHAT_MESSAGE */
const socketEmitNewUserMessage = ({ message, socket, username, roomName }: {
  message: string,
  socket?: SocketIOClient.Socket,
  username?: string, roomName?: string,
}): void => {
  if (!socket) return;
  socket.emit(SOCKETS.CHAT_MESSAGE, { username, message, roomName });
};

/* SOCKETS.PENALTY_ROWS */
const socketEmitPenaltyRows = ({ rowsNumber, username, roomName, socket }: {
  rowsNumber: number,
  username?: string,
  roomName?: string,
  socket?: SocketIOClient.Socket,
}): void => {
  if (!socket) return;
  console.log('PENALTY_ROWS emit, rowsNumber:', rowsNumber);
  socket.emit(SOCKETS.PENALTY_ROWS, { username, roomName, rowsNumber });
};

/* SOCKETS.UPDATE_SCORE */
const socketEmitUpdatePlayerScore = ({setCurrentPlayer, socket, isSoloMode }: {
  setCurrentPlayer: Dispatch<SetStateAction<PlayerType | undefined>>,
  isSoloMode: boolean,
  socket?: SocketIOClient.Socket,
}): void => {
  if (!socket) return;
  setCurrentPlayer((prevCurrentPlayer) => {
    if (!prevCurrentPlayer) return;
    socket.emit(SOCKETS.UPDATE_SCORE, { username: prevCurrentPlayer.username, roomName: prevCurrentPlayer.room, score: prevCurrentPlayer.score, isSoloMode });
    return prevCurrentPlayer;
  });
};

/* SOCKETS.GAMEOVER */
const socketEmitGameover = ({ isSoloMode, setCurrentPlayer, userContext, socket }: {
  isSoloMode: boolean,
  setCurrentPlayer: Dispatch<SetStateAction<PlayerType | undefined>>,
  userContext: UserContextType,
  socket?: SocketIOClient.Socket,
}): void => {
  if (isSoloMode) {
    return socketEmitUpdatePlayerScore({ setCurrentPlayer, isSoloMode, socket });
  } else {
    if (!socket) return;
    socket.emit(SOCKETS.GAMEOVER, { username: userContext.username, roomName: userContext.room });
  }
};

/* SOCKETS.UPDATE_SPECTRUM */
const socketEmitUpdateSpectrum = ({ spectrum, userContext, socket }: {
  spectrum: Matrix,
  userContext: UserContextType,
  socket?: SocketIOClient.Socket,
}): void => {
  if (!socket) return;
  socket.emit(SOCKETS.UPDATE_SPECTRUM, { username: userContext.username, roomName: userContext.room, spectrum });
};

/* SOCKETS.MORE_TETRIS_TILES */
const socketEmitMoreTetrisTiles = (userContext: UserContextType, socket?: SocketIOClient.Socket): void => {
  if (!socket) return;
  socket.emit(SOCKETS.MORE_TETRIS_TILES, { username: userContext.username, roomName: userContext.room });
};

/////////////////////////////////////////////////////////////////////////////////////////////////////////////

/*
** socket.on events:
*/

/* SOCKETS.CHAT_MESSAGE */
const socketReceiveChatMessage = (message: Message): void => {
  addResponseMessage(message.username + ': ' + message.text, message.username);
};

/* SOCKETS.FETCH_CURRENT_PLAYER */
const socketReceiveCurrentPlayer = ({player, setCurrentPlayer}: {
  player?: PlayerType,
  setCurrentPlayer: Dispatch<SetStateAction<PlayerType | undefined>>,
}): void => {
  setCurrentPlayer(player);
  console.log('FETCH_CURRENT_PLAYER', player);
};

/* SOCKETS.UPDATE_ROOM_PLAYERS */
const socketReceiveUpdateRoomPlayers = ({ setRoomPlayers, setRoomLeader, setCurrentPlayer, data }: {
    setRoomPlayers: Dispatch<SetStateAction<PlayerType[] | []>>,
    setRoomLeader: Dispatch<SetStateAction<PlayerType | undefined>>,
    setCurrentPlayer: Dispatch<SetStateAction<PlayerType | undefined>>,
    data: { room: string, players: PlayerType[] }
  }): void => {
  setRoomPlayers(data.players);
  // When old leader leaves the room, we set a new leader
  // And if new room leader and this room player are the same user, we update isLeader in player (or update the whole player)
  const newLeader = _.find(data.players, (player) => player.isLeader);
  setRoomLeader(newLeader);
  setCurrentPlayer((prevCurrentPlayer) => {
    if (prevCurrentPlayer && newLeader && prevCurrentPlayer.id === newLeader.id) {
      return newLeader;
    }
    return prevCurrentPlayer;
  });
};

/* SOCKETS.START_GAME */
const socketReceiveStartGame = ({ setTileStack, setGameStarted, setIsPause, tileStack }: {
  setTileStack: Dispatch<SetStateAction<TetriminosType[] | ['O']>>,
  setGameStarted: Dispatch<SetStateAction<boolean>>, 
  setIsPause: Dispatch<SetStateAction<boolean>>,
  tileStack: TetriminosType[],
}): void => {
  // TODO: assign tile stack here
  console.log('START_GAME', tileStack);
  setTileStack(tileStack);
  setGameStarted(true);
  setIsPause(false);
};

/* SOCKETS.PAUSE_GAME */
const socketReceivePauseGame = (setIsPause: Dispatch<SetStateAction<boolean>>): void => {
  setIsPause(prevState => !prevState);
};

/* SOCKETS.PENALTY_ROWS */
const socketReceivePenaltyRows = ({ setMatrix, rowsNumber }:
  { setMatrix: Dispatch<SetStateAction<Matrix>>, rowsNumber: number }): void => {
  setMatrix((prevMatrix) => {
    const newMatrix = addPenaltyRows(prevMatrix, rowsNumber);
    console.log('PENALTY_ROWS receive. rowsNumber, newMatrix:', rowsNumber, newMatrix);
    return newMatrix;
  });
};

/* SOCKETS.REDIRECT_TO_RANKING */
const socketReceiveRedirectToRanking = ({ username, room, navigation }: {
  username: string | undefined,
  room: string | undefined,
  navigation: StackNavigationProp<RootStackParamList, 'Playground'>,
}): void => {
  navigation.replace('Ranking', { username, room });
};

/* SOCKETS.GAMEOVER */
const socketReceiveGameover = ({ setCurrentPlayer, setIsPause, setMatrix, setGameover, isSoloMode, socket, data }: {
  setCurrentPlayer: Dispatch<SetStateAction<PlayerType | undefined>>,
  setIsPause: Dispatch<SetStateAction<boolean>>,
  setMatrix: Dispatch<SetStateAction<Matrix>>,
  setGameover: Dispatch<SetStateAction<boolean>>,
  isSoloMode: boolean,
  socket?: SocketIOClient.Socket,
  data: { players: PlayerType[], endGame: boolean },
}): void => {
  if (!socket) return;
  const roomWinner = _.find(data.players, (player) => player.isWinner);

  setCurrentPlayer((prevCurrentPlayer) => {
    if (prevCurrentPlayer && roomWinner && roomWinner.username === prevCurrentPlayer.username) {
      const playerWithUpdatedScore = { ...prevCurrentPlayer, score: prevCurrentPlayer.score + SCORING.LAST_PLAYER };
      return playerWithUpdatedScore;
    }
    return prevCurrentPlayer;
  });

  if (data.endGame) {
    socketEmitUpdatePlayerScore({ setCurrentPlayer, isSoloMode, socket });
    setIsPause(true);
    setMatrix(blankMatrix);
    setGameover(true);
  }
};

/* SOCKET.UPDATE_SPECTRUM */
const socketReceiveUpdateSpectrum = (roomPlayers: PlayerType[], setRoomPlayers: Dispatch<SetStateAction<PlayerType[] | []>>): void => {
  console.log('SOCKET.UPDATE_SPECTRUM, roomPlayers', roomPlayers);
  setRoomPlayers(roomPlayers);
};

/* SOCKETS.MORE_TETRIS_TILES */
const socketReceiveMoreTetrisTiles = (tileStack: TetriminosType[], setTileStack: Dispatch<SetStateAction<TetriminosType[] | ['O']>>): void => {
  setTileStack((prevTileStack) => [...prevTileStack, ...tileStack]);
};

/* SOCKETS.SPEED_MODE */
const socketReceiveSpeedMode = (setSpeedMode: Dispatch<SetStateAction<boolean>>): void => {
  setSpeedMode((prevSpeedMode) => !prevSpeedMode);
};

export {
  socketReceiveChatMessage,
  socketEmitNewUserMessage,
  socketReceiveCurrentPlayer,
  socketReceiveUpdateRoomPlayers,
  socketReceiveStartGame,
  socketReceivePauseGame,
  socketEmitPenaltyRows,
  socketReceivePenaltyRows,
  socketEmitUpdatePlayerScore,
  socketReceiveRedirectToRanking,
  socketEmitGameover,
  socketReceiveGameover,
  socketEmitUpdateSpectrum,
  socketReceiveUpdateSpectrum,
  socketEmitMoreTetrisTiles,
  socketReceiveMoreTetrisTiles,
  socketReceiveSpeedMode,
};
