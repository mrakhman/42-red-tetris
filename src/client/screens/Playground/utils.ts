import _ from 'lodash';

import { cellState } from '/client/constants/tetriminos';

const formatChatTitle = (leader: string): string => {
  return `Leader: ${leader}`;
};

const formatChatSubtitle = (players: string[]): string => {
  if (players.length === 0) return 'No players';
  return `Players: ${players.join(', ')}`;
};

const roomPlayersNames = (players?: PlayerType[]): string[] => {
  if (!players || players.length === 0) return [];
  const namesArray = players.map((player) => player.username);
  return namesArray;
};

const convertMatrixToSpectrum = (matrix: Matrix): Matrix => {
  const result = _.cloneDeep(matrix);
  for (let j = 0; j < matrix[0].length; j++) {
    let flag = false;
    for (let i = 0; i < matrix.length; i++) {
      if (matrix[i][j] === cellState.OCCUPIED)
        flag = true;
      if (flag)
        result[i][j] = cellState.OCCUPIED;
    }
  }
  return result;
};

const filteredOpponents = (roomPlayers: PlayerType[], currentPlayerUsername: string): PlayerType[] => {
  return _.filter(roomPlayers, (player) => player.username !== currentPlayerUsername);
};

export { formatChatSubtitle, formatChatTitle, roomPlayersNames, convertMatrixToSpectrum, filteredOpponents };
