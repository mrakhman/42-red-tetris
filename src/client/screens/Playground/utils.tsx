const formatChatSubtitle = (players: string[]): string => {
  if (players.length === 0) return 'No players';
  return `Players: ${players.join(', ')}`;
};

export default formatChatSubtitle;
