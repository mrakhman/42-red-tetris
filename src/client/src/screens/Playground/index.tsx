import _ from 'lodash';
import { View, Text } from 'react-native';
import * as React from 'react';
import useInterval from '@use-it/interval';
import useKey from 'use-key-hook';

import { blankMatrix, blockMatrix } from '/constants/tetriminos';
import { blockType } from '/constants/tetriminos';
import { keyboard } from '/constants/keyboard';
import Block from '/models/block';
import Matrix from '/components/Matrix';
import Gameboy from '/components/Gameboy';
import Digits from '/components/Digits';

const printBlock = (matrix: Matrix, block: Block) => { 
  const { shape, xy } = block;
  shape.forEach((m, i) => (
    m.forEach((n, j) => {
      if (xy[0] + i >= 0 && xy[0] + i < 20 && xy[1] + j < 10) {
        matrix[xy[0] + i][xy[1] + j] = shape[i][j];
      }
    })
  ));
  return matrix;
};

export default function Playground(): JSX.Element {
  const [block, setBlock] = React.useState<Block>(new Block({
    type: _.sample<TetriminosType>(blockType) ?? 'T',
  }));
  const nextBlockType = blockType[(_.indexOf(blockType, block.type) + 1) % _.size(blockType)];
  useInterval(() => {
    setBlock((currentBlock) =>
      (currentBlock.xy[0] + currentBlock.shape.length < 20)
        ? currentBlock.fall()
        : new Block({ type: nextBlockType }));
  }, 500);
  useKey((key: number) => {
    if (key === keyboard.rotate) setBlock((currentBlock) => currentBlock.rotate());
    if (key === keyboard.left) setBlock((currentBlock) => currentBlock.left());
    if (key === keyboard.right) setBlock((currentBlock) => currentBlock.right());
    if (key === keyboard.down) setBlock((currentBlock) => currentBlock.fall());
  });
  const newMatrix = printBlock(blankMatrix(), block);
  return (
    <Gameboy>
      <View style={{ flexDirection: 'row', alignSelf:'flex-start', width: '100%' }}>
        <Matrix matrix={newMatrix} />
        <View style={{ marginLeft: 20, flex: 1 }} >
          <Text style={{ fontSize: 20 }}>Score</Text>
          <Digits style={{ marginVertical: 10, alignSelf: 'flex-end' }} />
          <Text style={{ fontSize: 20 }}>Start Line</Text>
          <Digits style={{ marginVertical: 10, alignSelf: 'flex-end' }} />
          <Text style={{ fontSize: 20 }}>Level</Text>
          <Digits style={{ marginVertical: 10, alignSelf: 'flex-end' }} />
          <Text style={{ fontSize: 20 }}>Next</Text>
          <Matrix
            matrix={_.merge(blockMatrix(), new Block({ type: nextBlockType }).shape)}
            style={{ borderWidth: 0, marginVertical: 10, alignSelf: 'flex-end' }}
          />
        </View>
      </View>
    </Gameboy>
  );
}
