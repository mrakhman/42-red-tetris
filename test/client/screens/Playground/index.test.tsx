import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Playground from '/client/screens/Playground/index';

describe.skip('Snapshots', () => {
  it.skip('should render Playground', () => {
    const wrapper = shallow(<Playground />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
