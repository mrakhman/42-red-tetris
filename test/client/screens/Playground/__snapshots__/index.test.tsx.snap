// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should render Playground 1`] = `
<Fragment>
  <Gameboy
    gameover={false}
    isPause={true}
    isSoloMode={true}
    roomPlayers={Array []}
  >
    <View
      style={
        Object {
          "alignSelf": "flex-start",
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <Matrix
        block={
          Object {
            "pos": Array [
              -2,
              4,
            ],
            "shape": Array [
              Array [
                0,
                0,
                1,
              ],
              Array [
                1,
                1,
                1,
              ],
              Array [
                0,
                0,
                0,
              ],
            ],
            "timeStamp": 1618306144971,
            "type": "L",
          }
        }
        matrix={
          Array [
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
          ]
        }
      />
      <View
        style={
          Object {
            "flex": 1,
            "marginLeft": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Score
        </Text>
        <Digits
          style={
            Object {
              "alignSelf": "flex-end",
              "marginVertical": 10,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Start Line
        </Text>
        <Digits
          style={
            Object {
              "alignSelf": "flex-end",
              "marginVertical": 10,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Level
        </Text>
        <Digits
          style={
            Object {
              "alignSelf": "flex-end",
              "marginVertical": 10,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Next
        </Text>
        <Matrix
          block={
            Object {
              "pos": Array [
                0,
                0,
              ],
              "shape": Array [
                Array [
                  0,
                  0,
                  0,
                  0,
                ],
                Array [
                  1,
                  1,
                  1,
                  1,
                ],
                Array [
                  0,
                  0,
                  0,
                  0,
                ],
                Array [
                  0,
                  0,
                  0,
                  0,
                ],
              ],
              "timeStamp": 1618306146194,
              "type": "I",
            }
          }
          matrix={
            Array [
              Array [
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
              ],
            ]
          }
          style={
            Object {
              "alignSelf": "flex-end",
              "borderWidth": 0,
              "marginVertical": 10,
            }
          }
        />
      </View>
    </View>
  </Gameboy>
</Fragment>
`;
